src/shared/parse-helpers.c

```diff
@@ -196,3 +198,53 @@ int parse_socket_bind_item(
         *port_min = mn;
         return 0;
 }
+
+int open_file_parse(const char *v, OpenFile **ret) {
+        _cleanup_free_ char *path = NULL, *fdname = NULL, *flags = NULL;
+        _cleanup_(open_file_freep) OpenFile *of = NULL;
+        int r;
+
+        assert(v);
+
+        of = new0(OpenFile, 1);
+        if (!of)
+                return log_oom();
+
+        r = extract_many_words(&v, ":", EXTRACT_DONT_COALESCE_SEPARATORS, &of->path, &of->fdname, &flags, NULL);
+        if (r < 0)
+                return r;
+
+        if (r == 0)
+                return -EINVAL;
+
+        if (!path_is_absolute(of->path))
+                return -EINVAL;
+
+        if (isempty(of->fdname)) {
+                r = free_and_strdup(&of->fdname, of->path);
+                if (r < 0)
+                        return r;
+        }
+
+        if (flags && streq(flags, "ro"))
+                of->flags = OPENFILE_RDONLY;
+
+        open_file_free_and_replace(*ret, of);
```

Comment:
Like I mentioned in our chat, we don't want to introduce this pattern incrementally because callers will have to check each function to see whether it uses `free_and_replace()` or `TAKE_PTR()` so let's use `TAKE_PTR()` here as well.